storage/src/test/java/org/apache/kafka/storage/internals/log/RemoteIndexCacheTest.java

```diff
@@ -0,0 +1,1207 @@
+        assertTrue(getIndexFileFromDisk(LogFileUtils.INDEX_FILE_SUFFIX).isPresent(), "Offset index file should be present on disk at " + tpDir.toPath());
+        assertTrue(getIndexFileFromDisk(LogFileUtils.TXN_INDEX_FILE_SUFFIX).isPresent(), "Txn index file should be present on disk at " + tpDir.toPath());
+        assertTrue(getIndexFileFromDisk(LogFileUtils.TIME_INDEX_FILE_SUFFIX).isPresent(), "Time index file should be present on disk at " + tpDir.toPath());
+
+        // add the spied entry into the cache, it will overwrite the non-spied entry
+        cache.internalCache().put(internalIndexKey, cacheEntry);
+
+        // no expired entries yet
+        assertEquals(0, cache.expiredIndexes().size(), "expiredIndex queue should be zero at start of test");
+
+        // call remove function to mark the entry for removal
+        cache.remove(internalIndexKey);
+
+        // wait until entry is marked for deletion
+        TestUtils.waitForCondition(cacheEntry::isMarkedForCleanup,
+                "Failed to mark cache entry for cleanup after invalidation");
+        TestUtils.waitForCondition(cacheEntry::isCleanStarted,
+                "Failed to cleanup cache entry after invalidation");
+
+        // first it will be marked for cleanup, second time markForCleanup is called when cleanup() is called
+        verify(cacheEntry, times(2)).markForCleanup();
+        // after that async it will be cleaned up
+        verify(cacheEntry).cleanup();
+
+        // verify that index(s) rename is only called 1 time
+        verify(cacheEntry.timeIndex()).renameTo(any(File.class));
+        verify(cacheEntry.offsetIndex()).renameTo(any(File.class));
+        verify(cacheEntry.txnIndex()).renameTo(any(File.class));
+
+        // verify no index files on disk
+        assertFalse(getIndexFileFromRemoteCacheDir(cache, LogFileUtils.INDEX_FILE_SUFFIX).isPresent(),
+                "Offset index file should not be present on disk at " + tpDir.toPath());
+        assertFalse(getIndexFileFromRemoteCacheDir(cache, LogFileUtils.TXN_INDEX_FILE_SUFFIX).isPresent(),
+                "Txn index file should not be present on disk at " + tpDir.toPath());
+        assertFalse(getIndexFileFromRemoteCacheDir(cache, LogFileUtils.TIME_INDEX_FILE_SUFFIX).isPresent(),
+                "Time index file should not be present on disk at " + tpDir.toPath());
+        assertFalse(getIndexFileFromRemoteCacheDir(cache, LogFileUtils.DELETED_FILE_SUFFIX).isPresent(),
+                "Index file marked for deletion should not be present on disk at " + tpDir.toPath());
+    }
+
+    private Optional<Path> getIndexFileFromDisk(String suffix) throws IOException {
+        return Files.walk(tpDir.toPath())
+                .filter(Files::isRegularFile)
+                .filter(path -> path.getFileName().toString().endsWith(suffix))
+                .findAny();
+    }
+
+    @Test
+    public void testCleanerThreadShutdown() throws IOException, InterruptedException {
```

Comment:
The logic in this test is a bit different than the original. From my testing the original test did not actually threw the `RuntimeException` to validate the cleaner thread could handle it as `cleanup()` is effectively only called from `remove()` and not by invalidating the entry in the underlying cache.